File: Python/src/kargonomi/webhook.py
from __future__ import annotations

import base64
import hashlib
import hmac
from typing import Literal


def verify_webhook_signature(
    raw_body: bytes,
    signature: str,
    secret: str,
    encoding: Literal["hex", "base64"] = "hex",
) -> bool:
    """Verify HMAC-SHA256 against the exact request bytes before parsing JSON."""
    digest = hmac.new(secret.encode(), raw_body, hashlib.sha256).digest()
    expected = digest.hex() if encoding == "hex" else base64.b64encode(digest).decode("ascii")
    return hmac.compare_digest(expected, signature.strip())

File: NodeJS/src/webhook.ts
import { createHmac, timingSafeEqual } from 'node:crypto';

export type WebhookSignatureEncoding = 'hex' | 'base64';

export function verifyWebhookSignature(rawBody: Uint8Array, signature: string, secret: Uint8Array, encoding: WebhookSignatureEncoding): boolean {
  if (signature.trim() === '' || secret.byteLength === 0) return false;
  let supplied: Buffer;
  try { supplied = Buffer.from(signature, encoding); }
  catch { return false; }
  if ((encoding === 'hex' && !/^[a-f\d]+$/iu.test(signature)) || supplied.length !== 32) return false;
  const expected = createHmac('sha256', secret).update(rawBody).digest();
  return supplied.length === expected.length && timingSafeEqual(supplied, expected);
}
